### ISSUE TITLE
Return a Future from app.serve

### ISSUE THREAD

**Issue by yoshuawuyts** [rustasync/tide#163]
## Detailed Description
Instead of creating a runtime inside `app.serve`, I propose we return a future instead that allows people to run the app on whichever runtime they choose.

## Context
From the API it currently isn't clear that Tide operates asynchronously. Also things like starting a database in parallel to Tide feels a bit awkward because async control flow primitives such as `try_join` and `select` don't work.

I propose we move from this:
```rust
#![feature(async_await)]

fn main() -> Result<(), failure::Error> {
    let mut app = tide::App::new(());
    app.at("/").get(async move |_| "Hello, world!");
    app.serve("127.0.0.1:8000")?;
}
```
To this:
```rust
#![feature(async_await)]

fn main() -> Result<(), failure::Error> {
    my_runtime::run(async {
        let mut app = tide::App::new(());
        app.at("/").get(async move |_| "Hello, world!");
        await!(app.serve("127.0.0.1:8000")?);
    })
}
```
We then shift the problem to making `my_runtime::run` nicer to use, but that's something that's worth solving for the whole async ecosystem, and not just Tide.

## Possible Implementation
We would need to patch this line: https://github.com/rustasync/http-service/blob/master/http-service-hyper/src/lib.rs#L94, and propagate it through to Tide.

**Comment by secretfader** [comment]
Within the bounds of this proposal, is it possible to still provide a "batteries-included" runtime experience for new users?

**Comment by LucioFranco** [comment]
@secretfader you could have it return a custom future, that then as a function on it call say `sync` that will block and start the server. Under the hood it would consume itself, pass itself to a runtime.

**Comment by Texlo-Dev** [comment]
I agree with this proposal. Currently, it is a challenge to set up an asynchronous database pool with Tide, as you can't use the same runtime as Tide.

**Comment by yoshuawuyts** [comment]
With [Runtime](https://github.com/rustasync/runtime) we'll be able to also get rid of most of the boilerplate this change would introduce 😊 :
```rust
#![feature(async_await)]

#[runtime::main(runtime_tokio::Tokio)]
fn main() -> Result<(), failure::Error> {
    let mut app = tide::App::new(());
    app.at("/").get(async move |_| "Hello, world!");
    await!(app.serve("127.0.0.1:8000"))?;
}

**Comment by tirr-c** [comment]
I'm working on this, the update on `http-service-hyper` is here: rustasync/http-service#21

**Comment by fairingrey** [comment]
Closed via #203
